tozny/resource_client_registration_token.go

```diff
@@ -0,0 +1,199 @@
+    return diag.FromErr(err)
+  }
+
+  tokenName := d.Get("name").(string)
+  tfAllowedTypes := d.Get("allowed_registration_client_types").([]interface{})
+  allowedTypes := []string{}
+  for _, tfAllowedType := range tfAllowedTypes {
+    allowedTypes = append(allowedTypes, tfAllowedType.(string))
+  }
+
+  createTokenResponse, err := toznySDK.CreateRegistrationToken(ctx, accountClient.CreateRegistrationTokenRequest{
+    AccountServiceToken: account.Token,
+    Name:                tokenName,
+    TokenPermissions: accountClient.TokenPermissions{
+      Enabled:      d.Get("enabled").(bool),
+      AllowedTypes: allowedTypes,
+    },
+  })
+
+  if err != nil {
+    return diag.FromErr(err)
+  }
+
+  clientRegistrationToken := createTokenResponse.Token
+  d.Set("token", clientRegistrationToken)
+
+  // Associate created token with Terraform state and signal success
+  d.SetId(fmt.Sprintf("%d%v", time.Now().Unix(), tokenName))
+
+  return diags
+}
+
+func resourceClientRegistrationTokenRead(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+  var diags diag.Diagnostics
+  var err error
+
+  toznySDK := m.(*e3db.ToznySDKV3)
+
+  toznyClientCredentialsFilePath := d.Get("client_credentials_filepath").(string)
+
+  if toznyClientCredentialsFilePath != "" {
+    toznySDK, err = e3db.GetSDKV3(toznyClientCredentialsFilePath)
+    if err != nil {
+      return diag.FromErr(err)
+    }
+  }
+
+  account, err := toznySDK.Login(ctx, toznySDK.AccountUsername, toznySDK.AccountPassword, "password", toznySDK.APIEndpoint)
+
+  if err != nil {
+    return diag.FromErr(err)
+  }
+
+  listedRegistrationTokens, err := toznySDK.ListRegistrationTokens(ctx, account.Token)
+
+  if err != nil {
+    return diag.FromErr(err)
+  }
+
+  var listed bool
+  for _, listedRegistrationToken := range *listedRegistrationTokens {
+    if listedRegistrationToken.Name == d.Get("name").(string) {
+      if listedRegistrationToken.Token == d.Get("token").(string) {
+        listed = true
+        d.Set("enabled", listedRegistrationToken.Permissions.Enabled)
+        d.Set("allowed_registration_client_types", []interface{}{listedRegistrationToken.Permissions.AllowedTypes})
+        break
+      }
+    }
+  }
+
+  if !listed {
+    d.Set("token", "")
+    d.Set("enabled", false)
+  }
+
+  return diags
+}
+
+func resourceClientRegistrationTokenDelete(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+  var diags diag.Diagnostics
+  var err error
+
+  toznySDK := m.(*e3db.ToznySDKV3)
+
+  toznyClientCredentialsFilePath := d.Get("client_credentials_filepath").(string)
+
+  if toznyClientCredentialsFilePath != "" {
+    toznySDK, err = e3db.GetSDKV3(toznyClientCredentialsFilePath)
+    if err != nil {
+      return diag.FromErr(err)
+    }
+  }
+
+  account, err := toznySDK.Login(ctx, toznySDK.AccountUsername, toznySDK.AccountPassword, "password", toznySDK.APIEndpoint)
+
+  if err != nil {
+    return diag.FromErr(err)
+  }
```

Comment:
I'm seeing essentially this same bit of code repeated in each method. I expect some repeating in Golang code due to its nature, but I'm wondering if this should be abstracted into a method. Eventually when we have better internal caching for clients, account info, tokens, etc. it would be nice to only need to change this code in a single method. Am I missing something that would make that a bad idea?

Reply:
nah, just that my quality bar for terraform code is only slightly higher than it is for test code, but excellent and accurate points so I will extract this out to a method